python/tests/unit/arcticdb/version_store/test_sort_merge.py

```diff
@@ -222,50 +217,303 @@ def test_prune_previous(lmdb_library):
+            lib.sort_and_finalize_staged_data("sym", mode=mode)
+        assert "col_0" in str(exception_info.value)
+        assert "FLOAT" in str(exception_info.value)
+        assert "INT" in str(exception_info.value)
+
+    @pytest.mark.parametrize("mode", [StagedDataFinalizeMethod.APPEND, StagedDataFinalizeMethod.WRITE])
+    def test_type_mismatch_in_staged_segments_throws_with_promotoable_types(self, lmdb_library, mode):
+        lib = lmdb_library
+
+        df1 = pd.DataFrame({"col": np.array([1], dtype="int64")}, index=pd.DatetimeIndex([np.datetime64('2023-01-01')]))
+        lib.write("sym", df1, staged=True)
+
+        df2 = pd.DataFrame({"col": ["test"]}, index=pd.DatetimeIndex([np.datetime64('2023-01-02')]))
+        lib.write("sym", df2, staged=True)
+
+        with pytest.raises(SchemaException) as exception_info:
+            lib.sort_and_finalize_staged_data("sym", mode=mode)
+        assert "INT" in str(exception_info.value)
+        assert "UTF_DYNAMIC" in str(exception_info.value)
+
+    @pytest.mark.parametrize("mode", [StagedDataFinalizeMethod.APPEND, StagedDataFinalizeMethod.WRITE])
+    def test_staged_segments_cant_be_reordered(self, lmdb_library, mode):
+        lib = lmdb_library
+
+        df1 = pd.DataFrame({"col_0": [1], "col_1": ["test"], "col_2": [1.2]}, index=pd.DatetimeIndex([pd.Timestamp(2024, 1, 1)]))
+        lib.write("sym", df1, staged=True)
+        
+        df2 = pd.DataFrame({"col_1": ["asd"], "col_2": [2.5], "col_1": [2]}, index=pd.DatetimeIndex([pd.Timestamp(2024, 1, 2)]))
+        lib.write("sym", df2, staged=True)
+
+        with pytest.raises(SchemaException) as exception_info:
+            lib.sort_and_finalize_staged_data("sym", mode=mode)
+        assert "col_0" in str(exception_info.value)
+        assert "col_1" in str(exception_info.value)
+        assert "col_2" in str(exception_info.value)
+
+
+class TestStreamDescriptorMismatchOnFinalizeAppend:
+    def init_symbol(self, lib, sym):
+        df = pd.DataFrame({"col_0": np.array([1], "int32"), "col_1": [0.5], "col_2": ["val"]}, index=pd.DatetimeIndex([pd.Timestamp(2024, 1, 1)]))
+        lib.write(sym, df)
+
+    def test_cannot_append_column_subset(self, lmdb_library):
+        lib = lmdb_library
+        self.init_symbol(lib, "sym")
+        
+        df = pd.DataFrame({"col_0": [1]}, index=pd.DatetimeIndex([pd.Timestamp(2024, 1, 1)]))
+        lib.write("sym", df, staged=True)
+        with pytest.raises(SchemaException) as exception_info:
+            lib.sort_and_finalize_staged_data("sym", mode=StagedDataFinalizeMethod.APPEND)
+        assert "col_1" in str(exception_info.value)
+        assert "col_2" in str(exception_info.value)
+
+    def test_cannot_append_reordered_columns(self, lmdb_library):
+        lib = lmdb_library
+        self.init_symbol(lib, "sym")
+        
+        df = pd.DataFrame({"col_1": [1.4], "col_0": [5], "col_2": ["val"]}, index=pd.DatetimeIndex([pd.Timestamp(2024, 1, 1)]))
```

Comment:
On Linux, `col_0` will have dtype `int64` here, force it to match that in `init_symbol`

Reply:
Make sure all dtypes below are specified, so it is clear whether they are meant to match the `init_symbol` dtypes or not